azure-iot-device/azure/iot/device/iothub/pipeline/pipeline_stages_iothub_mqtt.py

```diff
@@ -55,11 +55,17 @@ def _execute_op(self, op):
                 ("api-version", pkg_constant.IOTHUB_API_VERSION),
                 ("DeviceClientType", pkg_constant.USER_AGENT),
             ]
-            username = "{hostname}/{client_id}/?{query_params}".format(
-                hostname=op.hostname,
-                client_id=client_id,
-                query_params=urllib.parse.urlencode(query_param_seq),
-            )
+            # According to the Azure IoT Hub User Agent Spec for MQTT:
+            # Append entire string to username field in CONNECT packet
+            # The spec also defines defines a custom user agent that the user can set.
+            if self.pipeline_root.pipeline_configuration.product_info:
+                username = self.pipeline_root.pipeline_configuration.product_info
```

Comment:
>username = self.pipeline_root.pipeline_configuration.product_info [](start = 16, length = 65)

This seems to replace the username, not append the `product_info` string to it.

Reply:
resolved in a rework.

---
In reply to: [337671382](https://github.com/Azure/azure-iot-sdk-python/pull/368#discussion_r337671382) [](ancestors = 337671382)